# Combining and Exporting Data

#### This Script Contains the Following Points:
#### 1. Importing Datasets to Jupyter
#### 2. Check Dimensions of Datasets
#### 3. Merge Datasets
#### 4. Check Merge Using Merge Flag
#### 5. Exporting Data

### 1. Importing Datasets to Jupyter

In [7]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [8]:
#importing data
path = r'/Users/kimkmiz/Documents/Instacart Basket Analysis 2024'

In [9]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'IC24 Prepared Data', 'orders_products_combined.pkl'))

In [10]:
# Import dataset products_clean.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'IC24 Prepared Data', 'products_clean.csv'), index_col = False)

### 2. Check the Dimensions of Datasets

In [12]:
# Check the output
df_ords_prods_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [13]:
df_ords_prods_combined.shape

(32434489, 10)

In [14]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [15]:
df_prods.shape

(49672, 6)

**Observations on Dimensions**
- df_prods not as wide and not as long
- unnamed: 0 column not needed
- merge column in df_ords_prod_combined not needed

In [17]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [18]:
#cell accidentally deleted, but 'unnamed:0' column was successfully dropped using drop function

In [19]:
#drop merge column from df_ords_prod_combined
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)

In [20]:
#check output
df_ords_prods_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

### 3. Merge Datasets

In [22]:
# Merge the updated dataframes
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [23]:
# Check the output
df_ords_prods_merge.head()

Unnamed: 0  product_id                product_name  aisle_id  \
0           0           1  Chocolate Sandwich Cookies        61   
1           0           1  Chocolate Sandwich Cookies        61   
2           0           1  Chocolate Sandwich Cookies        61   
3           0           1  Chocolate Sandwich Cookies        61   
4           0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  order_id  user_id  order_number  order_day_of_week  \
0             19     5.8   3139998      138            28                  6   
1             19     5.8   1977647      138            30                  6   
2             19     5.8    389851      709             2                  0   
3             19     5.8    652770      764             1                  3   
4             19     5.8   1813452      764             3                  4   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

### 4. Check Merge Using Merge Flag

In [25]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

**Observations**
- merged using inner join
- merged datasrt has 32404859 rows with information from both datasets

In [41]:
df_ords_prods_merge.shape

(32404859, 15)

In [45]:
df_ords_prods_merge.describe()

Unnamed: 0    product_id      aisle_id  department_id        prices  \
count  3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07   
mean   2.560037e+04  2.559866e+04  7.119612e+01   9.919792e+00  7.790180e+00   
std    1.408555e+04  1.408400e+04  3.821139e+01   6.281485e+00  4.242125e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00   
25%    1.354400e+04  1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00   
50%    2.530300e+04  2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00   
75%    3.795100e+04  3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01   
max    4.969200e+04  4.968800e+04  1.340000e+02   2.100000e+01  2.500000e+01   

           order_id       user_id  order_number  order_day_of_week  \
count  3.240486e+07  3.240486e+07  3.240486e+07       3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01       2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01       2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00       1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01       3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01       6.000000e+00   

       order_hour_of_day  days_since_prior_order  add_to_cart_order  \
count       3.240486e+07            3.032876e+07       3.240486e+07   
mean        1.342515e+01            1.110408e+01       8.352547e+00   
std         4.246380e+00            8.779064e+00       7.127071e+00   
min         0.000000e+00            0.000000e+00       1.000000e+00   
25%         1.000000e+01            5.000000e+00       3.000000e+00   
50%         1.300000e+01            8.000000e+00       6.000000e+00   
75%         1.600000e+01            1.500000e+01       1.100000e+01   
max         2.300000e+01            3.000000e+01       1.450000e+02   

          reordered  
count  3.240486e+07  
mean   5.895873e-01  
std    4.919087e-01  
min    0.000000e+00  
25%    0.000000e+00  
50%    1.000000e+00  
75%    1.000000e+00  
max    1.000000e+00

### 5. Exporting Data

In [26]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','IC24 Prepared Data', 'ords_prods_merge.pkl'))